Fixes #30862 - introduce SettingRegistry

ezr-ondrej opened the pull request:
SettingRegistry keeps all the setting information as set of SettingPresenters.
This registry should be the public API for accessing setting values.

The public interface should be called as `Foreman.settings[<name>] => value`

This allows new DSL that will be introduced in: #8437
And can be followed by serving these on index pages: #8438

ezr-ondrej committed:
Fixes #30862 - introduce SettingRegistry

SettingRegistry keeps all the setting information as set of SettingPresenters.
This registry should be the public API for accessing setting values.

The public interface should be called as

`Foreman.settings[<name>] => value`

ares commented:
This (and related PRs) would greatly benefit from better description of the change. Ideally provide what each PR changes and how to test that. Bonus for also mentioning the motivation for these changes.

tbrisker commented on lib/foreman.rb:
```diff
@@ -30,4 +30,8 @@ def self.instance_id
   end
+
+  def self.settings
+    @settings ||= SettingRegistry.new
```
Just thinking out loud, could this cause multiple registries to be created in multi-threaded environments?

tbrisker commented on app/registries/setting_registry.rb:
```diff
@@ -0,0 +1,30 @@
+class SettingRegistry
```
This folder is only loaded once during initialization. see https://github.com/theforeman/foreman/blob/develop/app/registries/README.md and also somewhat old https://projects.theforeman.org/projects/foreman/wiki/Registry_pattern - just to make sure that you've taken into account the considerations regarding reloading in development and impact on tests

ezr-ondrej replied:
I'd keep it simple for now, as it is used during the initialization process, I've used simple singleton
Later I'd like to hava lazy initialization and it could be thread safe, every instance would load the settings from the Setting classes itself.

tbrisker commented on app/registries/setting_registry.rb:
```diff
@@ -0,0 +1,30 @@
+    raise "setting name #{name} is not unique!" if @settings.key?(name.to_s)
+    @settings[name.to_s] = SettingPresenter.new({ :category => category,
+                                                  :name => name,
+                                                  :preset_value => value,
```
What is `preset_value` for and how does it differ from `default`?

ezr-ondrej replied:
hmm now you're mentioning it, it should probably be saved to the DB, as it is the first value we set for some settings, e.g. `instance_id`

ezr-ondrej replied:
It's saving the value to the database, if its not there yet.

tbrisker commented on app/controllers/api/v2/settings_controller.rb:
```diff
@@ -9,6 +9,7 @@ class SettingsController < V2::BaseController
 
       def index
         @settings = resource_scope_for_index.live_descendants
+        @settings.map! { |s| Foreman.settings.find(s.name) }
```
`@settings` is a relation here, so `map!` probably doesn't work on it:
```suggestion
        @settings = resource_scope_for_index.live_descendants.map { |s| Foreman.settings.find(s.name) }
```
Although, shouldn't we pull the settings from the registry rather than the db here?

ezr-ondrej replied:
We should in the future, but I'm trying to keep the changes minimal here and getting it from registry would require implementing the search on the registry or dropping the search, I'd like to do so in a followup PR, when we already have the registry implementation in.

ezr-ondrej replied:
I guess I missunderstood this comment, fixed it according to your comment.

ezr-ondrej commented:
[test foreman]

ezr-ondrej commented on app/controllers/concerns/application_shared.rb:
```diff
@@ -19,6 +19,10 @@ def set_current_user(user)
   end
 
+  def load_settings
+    Foreman.settings.load
```
This reloads the definitions and values from database on every request.

ezr-ondrej commented on app/models/setting.rb:
```diff
@@ -94,17 +94,11 @@ def self.cache_key(name)
-    cache.fetch(cache_key(name)) do
-      find_by(:name => name)&.value
-    end
+    Foreman.settings[name]
```
We are dropping `cache` here!

ezr-ondrej replied:
I've removed the cache entirely

ezr-ondrej commented:
Last katello failing tests are just about Katello updating the settings directly in database and expect the changes to be instantly availabe in test, but this is never necessary, IRL there would be another request where the setting would get reloaded, but I've filed a PR to fix this to use public API and then the change should be seen immediatelly - thus the failures would be fixed: #9272

ezr-ondrej commented on app/presenters/setting_presenter.rb:
```diff
@@ -51,6 +45,13 @@ def id
     name
   end
 
+  def parse_and_update_value(value)
+    Foreman::Logging.logger('app').warn("Setting #{name} has no database value and can not be updated") and return unless db_model
+    db_model.parse_string_value(value)
+    res = db_model.save
+    res
+  end
```
I did not figure how to get rid of this method, because we need something to proxy the errors in `Api::V2::SettingsController#update`

tbrisker replied:
How about calling something like `Foreman.settings.update(name, value)` directly in the controller? Probably the parsing of string values should be done in the registry as well if the value is a string, and you would get the same errors

ezr-ondrej commented:
[test foreman][test upgrade]

ezr-ondrej commented:
[test katello]

ezr-ondrej commented:
[test upgrade]

tbrisker commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,96 @@
+    find(name)&.value
+  end
+
+  def update(name, value)
```
to be consistent, maybe this should be `[]=`?

ezr-ondrej replied:
Yeah, I guess we can, I just didn't want to make it a public API and `[]=` looks much more as a public API :) but why not ;)

tbrisker replied:
why not make it public? long term I see the registry as the main point of interaction with settings, so that if you want to set a value the best practice would be using this method

tbrisker commented on app/presenters/setting_presenter.rb:
```diff
@@ -80,4 +81,8 @@ def category_label
     category.gsub(/Setting::/, '')
   end
+
+  def select_values
+    Setting.select_collection_registry.collection_for name
+  end
```
I guess in the future we could merge the select_collection_registry into the main setting registry?

ezr-ondrej replied:
Definitelly, it actually got me started to think about this in the first place, so it is in the roadmap xD

tbrisker commented on app/presenters/setting_presenter.rb:
```diff
@@ -7,38 +7,32 @@ class SettingPresenter
     Setting.model_name
   end
 
+  # This refreshes a value for the entry from db
+  def load_db_value(db_model)
+    @db_model = db_model
+    if db_model
+      v = db_model[:value]
+      v = db_model.decrypt_field(v)
+    end
+    self.value = v.nil? ? default : YAML.safe_load(v)
+  end
```
Thinking more about this... we could keep it only in the model and set the value directly according to what we get from the model. It might be that it's the default, but that's ok. It would make loading simpler as well (see next comment)

ezr-ondrej replied:
yeah, make sense, the serialization should be left in the data layer, so it will be much cleaner :+1:

tbrisker commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,96 @@
+    db_record = _find_or_new_db_record(name)
+    db_record.value = value
+    db_record.save!
+    definition.load_db_value(db_record)
```
and then this becomes 
```suggestion
    definition.value = db_record.value
```

tbrisker commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,96 @@
+      set_definition = find(set.name)
+      logger.warn("Setting #{set.name} has no definition, clean up your database") and next unless set_definition
+      loaded_names << set.name
+      set_definition.load_db_value(set)
```
this would be 
```suggestion
      set_definition.value = set.value
```

tbrisker commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,96 @@
+      set_definition.load_db_value(set)
+    end
+    # load nil to set value to default
+    (@settings.keys - loaded_names).each do |name|
+      find(name).load_db_value(nil)
+    end
```
And this
```suggestion
    @settings.except(loaded_names).each do |setting|
      setting.value = setting.default
    end
```

tbrisker commented on app/presenters/setting_presenter.rb:
```diff
@@ -7,33 +7,16 @@ class SettingPresenter
+  attribute :config_file, :string
+  attribute :updated_at
+
+  attr_reader :db_model
```
I guess this is not needed anymore?

tbrisker commented on app/presenters/setting_presenter.rb:
```diff
@@ -7,33 +7,14 @@ class SettingPresenter
   attribute :category, :string, default: 'Setting::General'
   attribute :name, :string
-  attribute :default
   attribute :value
+  attribute :default
```
I'm guessing this order change is by accident, can you revert it for a cleaner history?

tbrisker commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,98 @@
+  #   Foreman.settings[:default_locale] => nil.
+  #
+  def [](name)
+    find(name)&.value
```
Should we raise an exception if there is no such setting? Or at least log it?

ezr-ondrej replied:
We can log it, exception would make sense, but there are :dragon:s and it would change from previous implementation.

tbrisker commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,98 @@
+
+  def []=(name, value)
+    definition = find(name)
+    raise "Setting definition for #{name} not found, can not set" unless definition
```
Should this be a `Foreman::Exception`?

ezr-ondrej commented:
[test upgrade]

tbrisker commented on test/mailers/application_mailer_test.rb:
```diff
@@ -34,7 +34,7 @@ def mail(to = 'nobody@example.com', subject = 'Danger, Will Robinson!')
 
   test 'foreman server header is set' do
-    assert_equal mail.header['X-Foreman-Server'].to_s, 'foreman.some.host.fqdn'
+    assert_equal mail.header['X-Foreman-Server'].to_s, SETTINGS[:fqdn]
```
This seems wrong:
https://github.com/theforeman/foreman/blob/develop/app/mailers/application_mailer.rb#L13

ezr-ondrej replied:
I was just missing asterisk at except, so you were right, I was trying to solve a wrong bug xD

ezr-ondrej commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,100 @@
+      set_definition.value = set.value
+    end
+    # load nil to set value to default
+    @settings.except(*loaded_names).each do |name, set_definition|
```
freaking asterisk

tbrisker replied:
sorry, my bad, i suggested this change :)

ezr-ondrej replied:
And I've blindly applied it without thinking, not really your bad :)
I remember struggling with exactly this few times already, I'm even thinking if the method should warn about that xD

ezr-ondrej commented:
:tada:? xD

tbrisker commented on app/models/setting.rb:
```diff
@@ -353,18 +339,11 @@ def clear_value_when_default
     readonly! if !new_record? && has_readonly_value?
   end
+
+  def reload_definition_value
```
i think this should be named `refresh_registry_value`

tbrisker commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,100 @@
+    # add() all setting definitions
+    Setting.descendants.each do |cat_cls|
+      if cat_cls.default_settings.empty?
+        # Setting category uses really old way of doin things
```
```suggestion
        # Setting category uses really old way of doing things
```

tbrisker commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,100 @@
+
+  def load
+    # add() all setting definitions
+    Setting.descendants.each do |cat_cls|
```
Looks like when starting the app this runs before core settings are loaded (meaning the `Setting.descendants` doesn't include all the subclassess yet), because we extend ActiveRecord with a method that calls `Setting[:per_page]` which triggers the singleton initializer:
https://github.com/theforeman/foreman/blob/develop/lib/core_extensions.rb#L119

Which in turn leads to logging a lot of messages, meaning we are using settings before the registery is fully loaded and initialized.

ezr-ondrej replied:
yes we are and it's known issue I believe, but there is not much we can do about it in this PR, in follow-up I think we should improve that and try not accessing settings too early.

ezr-ondrej replied:
I believe I've 'fixed' it, by manual initialization and by throwing a warning if setting is used too soon

tbrisker reviewed (changes requested):
Good thing you added logging when there's some conditions that shouldn't occur, that led to finding where the singleton actually got initialized (hint: earlier than expected)

ezr-ondrej committed:
Refs #30862 - SettingRegistry lazy initialization

Initialize the registry at specific moment and note if the registry is
ready or not.

ezr-ondrej commented:
I've included the proposal for early used setting as separate commit.

tbrisker commented on lib/core_extensions.rb:
```diff
@@ -115,7 +115,7 @@ def self.unconfigured?
   def self.per_page
     # Foreman.in_rake? prevents the failure of db:migrate for postgresql
     # don't query settings table if in rake
-    return 20 if Foreman.in_rake?
+    return 20 unless Foreman.settings.ready?
     Setting[:entries_per_page] rescue 20
   end
```
I believe we can actually get rid of this monkeypatch and just set
```
WillPaginate.per_page = Setting[:entries_per_page]
```
in the will_paginate initializer that runs after the settings have already been loaded.

ezr-ondrej replied:
That would prevent us from changing this value without restart of foreman though.

ezr-ondrej replied:
There would have to be an after_safe hook, that would set the value `WillPaginate.per_page` after the setting page and that is out of scope of this PR. I believe this is the same what we had before, we don't need pagination from setting in initialization process, so the result will be fine IMHO.

tbrisker replied:
This should probably move to `ApplicationRecord` which would fix the issue but doesn't have to be in the same pr.

ezr-ondrej committed:
Refs #30862 - Setting creation without validation

We don't need to validate the default setting created from code.

tbrisker reviewed (commented):
On my dev setup, these settings are in the db but don't have a definition: 
`use_gravatar trusted_puppetmaster_hosts legacy_puppet_hostname puppetrun max_trend enable_orchestration_on_fact_import ansible_become Parametrized_Classes_in_ENC Default_variables_Lookup_Path remote_addr foreman_tasks_proxy_action_start_timeout ansible_port ansible_user ansible_ssh_pass modulepath`
I think most cases are the outcome of a development db that has rolled migrations back and forth for years, but would be good to make sure all of them are actually cleaned from the db (otherwise users will get a bunch of annoying warnings on every request)

tbrisker commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,107 @@
+    loaded_names = []
+    Setting.unscoped.all.each do |set|
+      set_definition = find(set.name)
+      logger.warn("Setting #{set.name} has no definition, clean up your database") and next unless set_definition
```
This should probably be `logger.debug` since it's not impacting the user, it should be just devs that need to see it and clean up some unused settings

ezr-ondrej replied:
Done

ezr-ondrej commented:
I've rebased. I believe the cleanup can be done as follow-up

ekohl commented:
> use_gravatar

db/migrate/20171208113210_remove_use_gravatar_setting.rb

> trusted_puppetmaster_hosts

db/migrate/20180228132500_rename_trusted_hosts.rb

> legacy_puppet_hostname

db/migrate/20180705164601_remove_legacy_puppet_hostname_setting.rb

> puppetrun

db/migrate/20200602155700_drop_puppet_run.rb

And then I stopped :)

It would be interesting to check this with a production database. If it shows there, our migrations are broken.

ezr-ondrej commented:
> It would be interesting to check this with a production database. If it shows there, our migrations are broken.

I strongly believe it is only develop database issue, you have a PR, that already include the change, thus the migration run, but then you test a PR that doesn't include this change yet, and you are stuck with the setting forever, so it is very easy to get to such state in development, but I can't imagine the same happening for production :shrug: :) still, there might be some for which we forgot the migrations.

ekohl commented:
I'm with you on that, but just for completeness here's the list from my 2.4 production installation that's been running for a while:
<details><summary>Foreman 2.4 production setting names dump</summary>

```
 access_unattended_without_build
 administrator
 always_show_configuration_status
 append_domain_name_for_hosts
 authorize_login_delegation
 authorize_login_delegation_api
 authorize_login_delegation_auth_source_user_autocreate
 bcrypt_cost
 bmc_credentials_accessible
 clean_up_failed_deployment
 create_new_host_when_facts_are_uploaded
 create_new_host_when_report_is_uploaded
 db_pending_migration
 db_pending_seed
 default_global_registration_item
 default_locale
 default_location
 default_organization
 Default_parameters_Lookup_Path
 default_puppet_environment
 default_pxe_item_global
 default_pxe_item_local
 default_timezone
 delivery_method
 destroy_vm_on_host_delete
 dns_timeout
 email_reply_address
 email_subject_prefix
 enc_environment
 entries_per_page
 excluded_facts
 failed_login_attempts_limit
 fix_db_cache
 foreman_url
 global_iPXE
 global_PXEGrub
 global_PXEGrub2
 global_PXELinux
 host_owner
 host_power_status
 http_proxy
 http_proxy_except_list
 idle_timeout
 ignored_interface_identifiers
 ignore_facts_for_domain
 ignore_facts_for_operatingsystem
 ignore_puppet_facts_for_provisioning
 instance_id
 instance_title
 intermediate_ipxe_script
 interpolate_erb_in_parameters
 lab_features
 libvirt_default_console_address
 local_boot_iPXE
 local_boot_PXEGrub
 local_boot_PXEGrub2
 local_boot_PXELinux
 location_fact
 login_delegation_logout_url
 login_text
 manage_puppetca
 matchers_inheritance
 maximum_structured_facts
 max_trend
 monitoring_affect_global_status
 monitoring_create_action
 monitoring_delete_action
 name_generator_type
 oauth_active
 oauth_consumer_key
 oauth_consumer_secret
 oauth_map_users
 oidc_algorithm
 oidc_audience
 oidc_issuer
 oidc_jwks_url
 organization_fact
 outofsync_interval
 proxy_request_timeout
 puppet_interval
 puppet_out_of_sync_disabled
 query_local_nameservers
 require_ssl_smart_proxies
 restrict_registered_smart_proxies
 root_pass
 rss_enable
 rss_url
 safemode_render
 sendmail_arguments
 sendmail_location
 send_welcome_email
 smtp_address
 smtp_authentication
 smtp_domain
 smtp_enable_starttls_auto
 smtp_openssl_verify_mode
 smtp_password
 smtp_port
 smtp_user_name
 ssh_timeout
 ssl_ca_file
 ssl_certificate
 ssl_client_cert_env
 ssl_client_dn_env
 ssl_client_verify_env
 ssl_priv_key
 token_duration
 trusted_hosts
 unattended_url
 update_environment_from_facts
 update_hostgroup_from_facts
 update_ip_from_built_request
 update_subnets_from_facts
 use_shortname_for_vms
 use_uuid_for_certificates
 websockets_encrypt
 websockets_ssl_cert
 websockets_ssl_key
```
</details>

tbrisker reviewed (dismissed):
Thanks @ezr-ondrej , great step forward! @ekohl any final comments?

tbrisker commented:
[test foreman]

ekohl commented on app/helpers/settings_helper.rb:
```diff
@@ -10,6 +10,7 @@ def cat_label(category)
   def grouped_settings(settings)
     settings.each_with_object({}) do |setting, memo|
       hash = setting_to_hash(setting)
+      next unless hash
```
This made me wonder if it needs to return memo at the end, but it doesn't appear to.

Wouldn't it be shorter and equivalent to replace this whole method with:
```ruby
settings.map { |s| setting_to_hash(s) }.compact.group_by { |s| s[:category] }
```

ezr-ondrej replied:
no, `each_with_object` doesn't require to return the memo, we can update the method, although I didn't want to have too much intrusive changes here, would it be ok to do it here: https://github.com/theforeman/foreman/pull/8438/files#r618352022 ?

ekohl replied:
I'm fine with that.

ekohl commented on app/helpers/settings_helper.rb:
```diff
@@ -20,7 +21,8 @@ def grouped_settings(settings)
   def setting_to_hash(setting)
-    presenter = SettingPresenter.from_setting(setting)
+    presenter = Foreman.settings.find(setting.name)
+    logger.warn("Setting #{setting.name} doesn't exist anymore, clean up your database") and return nil if presenter.nil?
```
I'm not a fan of the `and return nil`. I prefer to be explicit here:
```suggestion
    if presenter.nil?
      logger.warn("Setting #{setting.name} doesn't exist anymore, clean up your database")
      return nil
    end
```

ekohl commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,107 @@
+    end
+  end
+
+  def _add(name, category:, default:, description:, full_name: nil, value: nil, encrypted: false)
```
Should there be a `private` before this:
```suggestion
  private

  def _add(name, category:, default:, description:, full_name: nil, value: nil, encrypted: false)
```

ezr-ondrej replied:
we are using it in tests, I prefer to have `_add`, but keep it public, so it is possible to use, if necessary, but known it is not public API

ekohl replied:
In tests you can use `.send(:add)` to call private methods, right? I think I'm fine either way. I come from a Python world where there are no real private methods, just underscore convention. I was mostly thinking about it because tools that generate API docs may not respect these "private" methods.

ezr-ondrej replied:
yeah, you can, though I like this more, I believe it is more readable, I didn't want to refuse access to this method, just hint it is not public API.
Should I add :nodoc: ?
It would be good idea, but we don't have documentation tool to target this comment to, so I'm not sure it makes sense.

ekohl replied:
Let's leave it for a documentation update then.

ekohl commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,107 @@
+  include Singleton
+
+  def initialize
+    @settings = {}
```
Just for my understanding (I like thinking in types) this is a `Hash[String, Setting]`, correct? I wonder if we should adopt yardoc style tags to have some documentation, but probably for another time.

ezr-ondrej replied:
It is `Hash[String, SettingPresenter]`.
I agree with you there, but we should first talk about what to document and what style to use on discourse.

ekohl replied:
Let's use this as input in that discussion.

ekohl commented on app/views/api/v2/settings/show.json.rabl:
```diff
@@ -1,3 +1,3 @@
-object SettingPresenter.from_setting(@setting)
+object Foreman.settings.find(@setting.name)
```
Why is this needed? Can `@setting` be `nil` and thus end up calling `nil.name`?

ezr-ondrej replied:
no, It would be caught by controller and responded 404 if @setting should be `nil`.
We are still operating with the model in controller, so this is the easiest how to be handling model in controller, but presenter for views.

ekohl commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,107 @@
+    loaded_names = []
+    Setting.unscoped.all.each do |set|
+      set_definition = find(set.name)
+      logger.debug("Setting #{set.name} has no definition, clean up your database") and next unless set_definition
```
Here I also think the `and next` part is way too subtle. Control flow should be obvious.
```suggestion
      unless set_definition
        logger.debug("Setting #{set.name} has no definition, clean up your database")
        next
      end
```

Then you can also take it up a step to make it even clearer:

```suggestion
      unless (set_definition = find(set.name))
        logger.debug("Setting #{set.name} has no definition, clean up your database")
        next
      end
```

ekohl commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,107 @@
+
+  def load_values
+    loaded_names = []
+    Setting.unscoped.all.each do |set|
```
Nit: I think `set` is odd because I think of a set as in a collection. I actually think `s` is a better name than `set`

ekohl commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,107 @@
+      set_definition = find(set.name)
+      logger.debug("Setting #{set.name} has no definition, clean up your database") and next unless set_definition
+      loaded_names << set.name
+      set_definition.updated_at = set.updated_at
+      set_definition.value = set.value
```
Shouldn't this check if the value changed and only then update it?
```suggestion
      next if set_definition.value == set.value

      set_definition.updated_at = set.updated_at
      set_definition.value = set.value
```

ezr-ondrej replied:
I don't believe so, I see no value in such if, we did load the model already, there is no reason not setting the value even if it is still the same, it is simple setter on a class, so it is probably even faster just setting it directly.

ekohl replied:
My reasoning is that you don't want to **always** set `updated_at`. Otherwise you may end up with useless values. Maybe that fear isn't really a problem.

ezr-ondrej replied:
I don't understand :thinking:
`set` is a setting model, so we just set the `updated_at` of the presenter to a value we have in database, so if the value did not change in the meantime, the `updated_at` timestamp will be the same, we just pass the value, so we don't keep the whole ActiveRecord model in memory just for proxying the `value` and `updated_at` methods to it.

ekohl replied:
Ah right, I was confused. I somehow assumed the definition was a model instance. It makes sense you don't want to keep the model reference.

ezr-ondrej committed:
Refs #30862 - SettingRegistry readability improvements

Improve guards readability.

ekohl commented on app/services/setting_registry.rb:
```diff
@@ -0,0 +1,109 @@
+        next
+      end
+      loaded_names << s.name
+      definition.updated_at = s.updated_at
```
https://github.com/theforeman/foreman/pull/8002#discussion_r618371563 is still a discussion. Highlighting it here so it doesn't get lost.

ekohl reviewed (approved):
I do wonder if we could summarize the classes and their relations in developer docs. Not a blocker to merge this (can also be written after stabilization). Do we want to squash the commits all together or how do you want to merge this?

ezr-ondrej commented:
I would love to have some documentation for this, I'll come up with something when the classes are stable.

I'd prefer squash merge here, the additional commits are mostly nits just for easier review, but keeping them is not bad eighter, they have their descriptions, so it won't make a mess in history - but squash would look better IMHO.

ekohl commented:
I think you're allowed to squash merge yourself. Please do so and correct the commit message to be correct.

ezr-ondrej commented:
Thanks @ekohl and @tbrisker for the reviews! :)

lzap commented:
For the record there is a follow-up commit: https://github.com/theforeman/foreman/pull/8464
